Tesla Model 3 / Model Y Car Port (Clean)

lukasloetkolben opened the pull request:
In a conversation with DebboR, we considered removing the Model S code from openpilot and adding a clean Model3 code instead. Here is an implementation with only Tesla Model 3/Y.

The best thing about this carport is that only a single bus had to be intercepted. 🎉
No external Panda is required, just a C3X and a harness box.

**Checklist Model 3**
- [x] added entry to CAR in selfdrive/car/tesla/values.py
- [x] test route added to [routes.py](https://github.com/commaai/openpilot/blob/master/selfdrive/car/tests/routes.py)
- [x] route with openpilot: b5f2b5b4ef23f4b6/00000008--d88c2210a5
- [x] route with stock system: b5f2b5b4ef23f4b6/00000009--f3a74d938c
- [x] car harness used (if comma doesn't sell it, put N/A): N/A

**Checklist Model Y**
- [x] added entry to CAR in selfdrive/car/tesla/values.py
- [x] test route added to [routes.py](https://github.com/commaai/openpilot/blob/master/selfdrive/car/tests/routes.py)
- [x] route with openpilot: 35f334fe58c5cdc5/00000089--c5ed862a40
- [x] route with stock system: 35f334fe58c5cdc5/00000001--d8d176b1af
- [x] car harness used (if comma doesn't sell it, put N/A): N/A


Tesla is in dashcam mode only, so Model 3 and Model Y will not yet appear in CARS.md or release notes.

**Additional Routes for the other Fingerprints:**
Model 3 2023 (RHD): 574d45dbc63e9fd8/00000000--0c80908e8c
Model 3 2023 (LHD): 8b67b428fd9fea5c/00000047--d04a4d7f8a
Model 3 2021 (LHD): 5df538ba7679f5b7/00000002--ab29e12a18
Model 3 2020 (LHD): be7997c2a9399899/00000001--9d87515889
Model Y 2022 (LHD): 58a1cdc98298efcb/00000000--166a842612
Model Y 2022 (RHD): 5313f747976147e6/00000015--c1fa5e2ce0
Model Y 2020 (LHD): 7588270b74f42d15/00000000--85ec85024a

**Demos on Youtube:**
https://www.youtube.com/watch?v=CwjCp1Gb9d4
https://www.youtube.com/watch?v=1rh0sfZPymE
https://www.youtube.com/watch?v=LC0k8hmkBus
https://www.youtube.com/watch?v=ljMDHocoxxM

lukasloetkolben committed:
WIP: Tesla Model 3

lukasloetkolben committed:
WIP: Tesla Model 3

lukasloetkolben committed:
bump panda

lukasloetkolben committed:
doorOpen to false for testing

lukasloetkolben committed:
- use EPAS3S_sysStatus instead of EPAS3P_sysStatus
- try to send cancel only once

lukasloetkolben committed:
bump opendbc

lukasloetkolben committed:
use EPAS3S_sysStatus at camera bus

lukasloetkolben committed:
bump panda

lukasloetkolben committed:
bump panda

lukasloetkolben committed:
remove epas3s

lukasloetkolben committed:
DAS_control added

lukasloetkolben committed:
EPAS3P to EPAS3S

lukasloetkolben committed:
bump panda

lukasloetkolben committed:
SCCM_rightStalk at can 1

lukasloetkolben committed:
wip model 3

lukasloetkolben committed:
remove external panda

lukasloetkolben committed:
Update interface.py

lukasloetkolben committed:
Update carstate.py

lukasloetkolben committed:
Update values.py

lukasloetkolben committed:
moved to EPAS3S steering angle

lukasloetkolben committed:
moved to EPAS3S steering angle

lukasloetkolben committed:
wip model 3

lukasloetkolben committed:
bump submodules

lukasloetkolben committed:
Update carstate.py

lukasloetkolben committed:
enable long control

lukasloetkolben committed:
bump panda

lukasloetkolben committed:
clean up && attempt to fix ACC get faulted

lukasloetkolben committed:
fixed keyerror in EPAS3S_sysStatus

lukasloetkolben committed:
Update carstate.py

lukasloetkolben committed:
new dbc

lukasloetkolben committed:
dbc update

lukasloetkolben committed:
testing long

lukasloetkolben committed:
testing long

lukasloetkolben committed:
Update carcontroller.py

lukasloetkolben committed:
Longitudinal control is working!

lukasloetkolben committed:
improve DAS_controlCounter

lukasloetkolben committed:
remove fingerprint

lukasloetkolben committed:
remove fingerprint

lukasloetkolben committed:
add fingerprints again

lukasloetkolben committed:
add fingerprints again

lukasloetkolben committed:
Update carcontroller.py

lukasloetkolben committed:
Update carcontroller.py

lukasloetkolben committed:
DAS_controlCounter fix

lukasloetkolben committed:
DAS_controlCounter fix

lukasloetkolben committed:
Update carcontroller.py

lukasloetkolben committed:
DAS_controlCounter simpler is better

lukasloetkolben committed:
try fw query

lukasloetkolben committed:
fingerprint model 3

lukasloetkolben committed:
fingerprint model 3

lukasloetkolben committed:
remove neutral gear bug

lukasloetkolben committed:
panda

lukasloetkolben committed:
Merge remote-tracking branch 'origin/model3' into model3

lukasloetkolben committed:
panda

lukasloetkolben committed:
panda

lukasloetkolben committed:
Update carcontroller.py

lukasloetkolben committed:
remove neutral gear bug

lukasloetkolben committed:
remove neutral gear bug

lukasloetkolben committed:
bump submodules

lukasloetkolben committed:
bump opendbc

lukasloetkolben committed:
vrabetz model3

lukasloetkolben committed:
cancel only in steering override if CS.cruiseState.enabled

lukasloetkolben committed:
Update carcontroller.py

lukasloetkolben committed:
sent cancel only if acc is enabled

lukasloetkolben committed:
cleanup

lukasloetkolben committed:
added doorOpen

lukasloetkolben committed:
bump opendbc

lukasloetkolben committed:
Cancel only if ACC is enabled

lukasloetkolben committed:
Cancel only if ACC is enabled

lukasloetkolben committed:
Update carstate.py

lukasloetkolben committed:
door state

lukasloetkolben committed:
VCFRONT_status 10Hz

lukasloetkolben committed:
bump panda

lukasloetkolben committed:
send neutral stalk position after canceling acc

lukasloetkolben committed:
bump panda

lukasloetkolben committed:
VCFRONT_status back to 50hz updated dbc

lukasloetkolben committed:
bump submodules

lukasloetkolben committed:
anyDoor open use different signal

lukasloetkolben committed:
bump submodules

lukasloetkolben committed:
VCLEFT_frontLatchSwitch

lukasloetkolben committed:
doorStatus on cp_adas

lukasloetkolben committed:
Merge remote-tracking branch 'origin/model3' into model3

# Conflicts:
#	selfdrive/car/tesla/carstate.py

lukasloetkolben committed:
Fingerprint Model3 2023 RHD (Por Por T)

lukasloetkolben committed:
bump opendbc

lukasloetkolben committed:
model3_cancel_acc counter bug fixed

lukasloetkolben committed:
cleanup

lukasloetkolben committed:
Fingerprint matu Model Y

lukasloetkolben committed:
commaai opendbc

lukasloetkolben committed:
Update .gitmodules

lukasloetkolben committed:
bump opendbc

lukasloetkolben committed:
Merge remote-tracking branch 'origin/master' into model3

# Conflicts:
#	panda
#	selfdrive/car/tesla/carstate.py
#	selfdrive/car/tesla/interface.py
#	selfdrive/car/tests/test_fw_fingerprint.py
#	selfdrive/car/torque_data/override.toml

lukasloetkolben committed:
bump submodules

lukasloetkolben committed:
added Model y

lukasloetkolben committed:
added Model y

lukasloetkolben committed:
bump panda

lukasloetkolben committed:
closer to master

lukasloetkolben committed:
bump panda

lukasloetkolben committed:
DAS_control frequency

lukasloetkolben committed:
bump panda

lukasloetkolben committed:
bump panda

lukasloetkolben committed:
Merge remote-tracking branch 'origin/model3' into model3

lukasloetkolben committed:
model3/y

lukasloetkolben committed:
bump submodules

lukasloetkolben committed:
model3/y

lukasloetkolben committed:
added torque data model3 / modely

lukasloetkolben committed:
bump panda

lukasloetkolben committed:
Merge remote-tracking branch 'origin/model3' into model3

lukasloetkolben committed:
added test routes

lukasloetkolben committed:
Merge remote-tracking branch 'origin/master' into model3

lukasloetkolben committed:
update submodules

lukasloetkolben committed:
update submodules

lukasloetkolben committed:
DAS_control at 25hz on model3/y

lukasloetkolben committed:
added snrubb model 3 2021 (atom) to fingerprints.py

lukasloetkolben committed:
added ursubpar model y 2023 to fingerprints.py

lukasloetkolben committed:
bump panda

lukasloetkolben committed:
Merge remote-tracking branch 'origin/model3' into model3

lukasloetkolben committed:
update panda

lukasloetkolben committed:
Merge remote-tracking branch 'origin/model3' into model3

lukasloetkolben committed:
added syamn model y 2022 to fingerprints.py

lukasloetkolben committed:
added syamn model y 2022 to fingerprints.py

lukasloetkolben committed:
Clean up Tesla code WIP

lukasloetkolben committed:
Clean up Tesla code WIP

lukasloetkolben committed:
WIP model3 clean

lukasloetkolben committed:
FwQueryConfig

lukasloetkolben committed:
cleanup model3

lukasloetkolben committed:
cleanup model3

lukasloetkolben committed:
switch panda to model3_clean

lukasloetkolben committed:
clean model3 wip

lukasloetkolben committed:
bump panda

lukasloetkolben committed:
dashcam only

lukasloetkolben committed:
Merge branch 'model3_clean' into model3_clean_pr

# Conflicts:
#	panda
#	selfdrive/car/tests/test_fw_fingerprint.py

lukasloetkolben committed:
commaai panda submodule

lukasloetkolben committed:
test_fw_query_timing

lukasloetkolben committed:
bump panda

lukasloetkolben committed:
updated test route

lukasloetkolben committed:
remove unused variable

lukasloetkolben committed:
Merge remote-tracking branch 'origin/model3_clean_pr' into model3_clean_pr

lukasloetkolben committed:
added TESLA_AP3_MODELY to fingerprints.py

lukasloetkolben committed:
Merge remote-tracking branch 'origin/model3_clean_pr' into model3_clean_pr

lukasloetkolben committed:
updated Model Y test route

lukasloetkolben committed:
merge master

lukasloetkolben committed:
added fingerprints for Model3 and ModelY

lukasloetkolben committed:
Update fingerprints.py

mattwebbio commented:
Is there anything preventing this from being merged?

lukasloetkolben committed:
ordered fingerprints

lukasloetkolben committed:
- blindspot signal
- same seatbelt signal for RHD & LHD

adeebshihadeh commented:
> Is there anything preventing this from being merged?

This is a substantially new car port, which takes time to review.

robbederks commented on selfdrive/car/tesla/carcontroller.py:
```diff
@@ -40,23 +41,24 @@ def update(self, CC, CS, now_nanos):
+    if self.frame % 10 == 0 and pcm_cancel_cmd and CS.acc_enabled:
+      counter = int(CS.sccm_right_stalk["SCCM_rightStalkCounter"] + 1) % 16
+      can_sends.append(self.tesla_can.right_stalk_press(counter, 1))  # half up (cancel acc)
+      can_sends.append(self.tesla_can.right_stalk_press((counter + 1) % 16, 0))  # to prevent neutral gear warning
```
little weird that it's `% 16` here again, let's make the initial counter variable be the actual current counter, and then `+1` and `+2` with the `% 16` on each line

robbederks commented on selfdrive/car/tesla/carstate.py:
```diff
@@ -29,78 +28,80 @@ def update(self, cp, cp_cam):
+    epas_status = cp.vl["EPAS3S_sysStatus"]
+    self.hands_on_level = epas_status["EPAS3S_handsOnLevel"]
+    self.steer_warning = self.can_define.dv["EPAS3S_sysStatus"]["EPAS3S_eacErrorCode"].get(int(epas_status["EPAS3S_eacErrorCode"]), None)
+    steer_status = self.can_define.dv["EPAS3S_sysStatus"]["EPAS3S_eacStatus"].get(int(epas_status["EPAS3S_eacStatus"]), None)
```
seems to be only used once, let's maybe inline

robbederks commented on selfdrive/car/tesla/carstate.py:
```diff
@@ -29,78 +28,80 @@ def update(self, cp, cp_cam):
 
     # Doors
-    ret.doorOpen = any((self.can_define.dv["GTW_carState"][door].get(int(cp.vl["GTW_carState"][door]), "OPEN") == "OPEN") for door in DOORS)
+    ret.doorOpen = any([cp_adas.vl["VCLEFT_doorStatus"]["VCLEFT_frontLatchSwitch"] != 1,
```
let's clean that up with list of tuples like `("VCLEFT_doorStatus", "VCLEFT_frontLatchSwitch", 1)`

robbederks replied:
maybe let's append the enums to the DBC like the model S port had

robbederks commented on selfdrive/car/tesla/carstate.py:
```diff
@@ -29,78 +28,80 @@ def update(self, cp, cp_cam):
-    self.acc_state = cp_cam.vl["DAS_control"]["DAS_accState"]
-    self.das_control_counters.extend(cp_cam.vl_all["DAS_control"]["DAS_controlCounter"])
+    self.sccm_right_stalk = copy.copy(cp_adas.vl["SCCM_rightStalk"])
+    self.das_control = copy.copy(cp_cam.vl["DAS_control"])
```
no need to copy the whole message, let's save just the counter val which is the only one used?

robbederks commented on selfdrive/car/tesla/teslacan.py:
```diff
@@ -86,9 +41,24 @@ def create_longitudinal_commands(self, acc_state, speed, min_accel, max_accel, c
+    return self.packer.make_can_msg("DAS_control", CANBUS.party, values)
+
+  def right_stalk_press(self, counter, position):
+    # TODO: Implement CRC checksum instead of lookup table.
```
would be _really_ nice to fix this

lukasloetkolben replied:
I have been working on this for a long time. I have not been able to work out the calculation using any method. I have tried CRC-beagle, revng. I tried to calculate it manually, but without success.

Perhaps there is a clever mind out there who can solve this?

robbederks replied:
@gregjhogan interested in trying? can upload a binary into ghidra if you want :)

dzid26 replied:
@lukasloetkolben  have you tried including message id before message bytes when doing reverse eng?

gregjhogan replied:
Here is the algorithm for`SCCM_rightStalk` tested on a few messages:
```python
import crcmod

msgs = [
  b'\x8c\x0b\x00',
  b'\x31\x0e\x00',
  b'\x44\x01\x00',
]

right_stalk_init_val = [0x7C, 0xB6, 0xF0, 0x2F, 0x69, 0xA3, 0xDD, 0x1C, 0x56, 0x90, 0xCA, 0x09, 0x43, 0x7D, 0xB7, 0xF1]

for msg in msgs:
  cntr = msg[1] & 0xF
  crc1_func = crcmod.mkCrcFun(0x12f, initCrc=0x00, xorOut=0xFF, rev=False)
  crc1 = crc1_func(msg[1:]) & 0xFF
  crc2_func = crcmod.mkCrcFun(0x12f, initCrc=crc1, xorOut=0xFF, rev=False)
  crc2 = crc2_func(bytes([right_stalk_init_val[cntr]])) & 0xFF
  print(f'message: 0x{msg.hex()} -> crc: {crc2:02X} = {"GOOD" if crc2 == msg[0] else "BAD"}')
```

lukasloetkolben replied:
just tested it on some more data. I will implement it later. thank you!

lukasloetkolben committed:
- changes requested by the review
- simplified doorOpen status

lukasloetkolben committed:
removed unused doorStatus signals

lukasloetkolben committed:
Cancel command can be sent directly, this will remove the neutral warning completely.

lukasloetkolben committed:
Revert "Cancel command can be sent directly, this will remove the neutral warning completely."

This reverts commit 290825dae4d2e679cc06db0ebba45c0549e28359.

lukasloetkolben committed:
Merge branch 'refs/heads/master' into model3_clean_pr

# Conflicts:
#	selfdrive/car/tests/test_fw_fingerprint.py

lukasloetkolben committed:
remove MODEL S comment

lukasloetkolben committed:
remove MODEL S comment

lukasloetkolben committed:
fingerprint jalsaeed model3 2020

lukasloetkolben committed:
implemented rightStalk crc, thanks to Greg for figuring it out.

jayvin100 commented:
If i have the 2024 M3 Highland will this work? I assume just have to add the fingerprint once I get it plugged in right?

lukasloetkolben commented:
> If i have the 2024 M3 Highland will this work? I assume just have to add the fingerprint once I get it plugged in right?

This is only for Model3 with HW3. The model3 Highland has HW4 at least has a different connector, I can't say anything about the software yet.

lukasloetkolben committed:
Merge remote-tracking branch 'origin/master' into model3_clean_pr

# Conflicts:
#	release/files_common

lukasloetkolben committed:
remove Model S from override.toml

lukasloetkolben committed:
radarUnavailable

lukasloetkolben committed:
- EPAS3S_eacStatus = status if active fault or not
- EPAS3S_eacErrorCode = last error code from current or previous fault

In order to tell if there is a fault right now, we need to look at eacStatus. However we need to ignore eacStatus when the fault is hands-on.

lukasloetkolben committed:
added ACCEL_MIN, ACCEL_MAX to CarControllerParams

lukasloetkolben committed:
Merge remote-tracking branch 'origin/master' into model3_clean_pr

# Conflicts:
#	selfdrive/car/tesla/carstate.py
#	selfdrive/car/tesla/radar_interface.py
#	selfdrive/car/tesla/teslacan.py

lukasloetkolben committed:
Merge branch 'refs/heads/master' into model3_clean_pr

# Conflicts:
#	selfdrive/car/fingerprints.py
#	selfdrive/car/tesla/carcontroller.py
#	selfdrive/car/tesla/carstate.py
#	selfdrive/car/tesla/fingerprints.py
#	selfdrive/car/tesla/interface.py
#	selfdrive/car/tesla/radar_interface.py
#	selfdrive/car/tesla/teslacan.py
#	selfdrive/car/tesla/values.py
#	selfdrive/car/tests/routes.py
#	selfdrive/car/tests/test_fw_fingerprint.py
#	selfdrive/car/torque_data/override.toml

lukasloetkolben committed:
add tesla
